recOrder/plugin/plugin_widget.py

```diff
@@ -0,0 +1,535 @@
+        else:
+            self.ui.le_bg_path.setText('Path Does Not Exist')
+
+    @pyqtSlot(bool)
+    def browse_acq_bg_path(self):
+        result = self._open_file_dialog(self.home_path)
+        self.acq_bg_directory = result
+        self.ui.le_bg_path.setText(result)
+
+    @pyqtSlot()
+    def enter_bg_correction(self):
+        state = self.ui.cb_bg_method.currentIndex()
+        if state == 0:
+            self.bg_option = 'None'
+        elif state == 1:
+            self.bg_option = 'global'
+        elif state == 2:
+            self.bg_option = 'local'
+
+    @pyqtSlot()
+    def enter_gpu_id(self):
+        self.gpu_id = self.ui.le_gpu_id
+
+    @pyqtSlot()
+    def enter_use_gpu(self):
+        state = self.ui.chb_use_gpu.checkState()
+        if state == 2:
+            self.use_gpu = True
+        elif state == 0:
+            self.use_gpu = False
+
+    @pyqtSlot()
+    def enter_obj_na(self):
+        self.obj_na = int(self.ui.le_obj_na.getText())
+
+    @pyqtSlot()
+    def enter_cond_na(self):
+        self.cond_na = int(self.ui.le_cond_na.getText())
+
+    @pyqtSlot()
+    def enter_mag(self):
+        self.mag = int(self.ui.le_mag.getText())
+
+    @pyqtSlot()
+    def enter_ps(self):
+        self.ps = int(self.ui.le_ps.getText())
+
+    @pyqtSlot()
+    def enter_n_media(self):
+        self.n_media = int(self.ui.le_n_media.getText())
+
+    @pyqtSlot()
+    def enter_pad_z(self):
+        self.pad_z = int(self.ui.le_pad_z.getText())
+
+    @pyqtSlot(bool)
+    def run_calibration(self):
+        logging.info('Starting Calibration')
+        self.ui.progress_bar.setValue(0)
+        self.calib.swing = self.swing
+        self.calib.wavelength = self.wavelength
+        self.calib.meta_file = os.path.join(self.directory, 'calibration_metadata.txt')
+        if self.calib.snap_manager.isLiveModeOn():
+            self.calib.snap_manager.setLiveModeOn(False)
+
+        self.calibration_thread = QThread()
+        self.calib_worker = CalibrationWorker(self, self.calib)
+        self.calib_worker.moveToThread(self.calibration_thread)
+        self.calibration_thread.started.connect(self.calib_worker.run)
+        self.calib_worker.finished.connect(self.calibration_thread.quit)
+        self.calib_worker.finished.connect(self.calib_worker.deleteLater)
+        self.calibration_thread.finished.connect(self.calibration_thread.deleteLater)
+        self.calib_worker.progress_update.connect(self.handle_progress_update)
+        self.calib_worker.extinction_update.connect(self.handle_extinction_update)
+        self.calib_worker.intensity_update.connect(self.handle_plot_update)
+        self.calib_worker.calib_assessment.connect(self.handle_calibration_assessment_update)
+        self.calib_worker.calib_assessment_msg.connect(self.handle_calibration_assessment_msg_update)
+        self.calibration_thread.start()
+        self._disable_buttons()
+        self.calibration_thread.finished.connect(self._enable_buttons)
```

Comment:
There's a certain pattern that might be useful here, but probably requires a refactor in a later PR:

Rather than create QThread every time a calibration call is made, you have a separate class (`QThreadWorker` for example) that either creates the thread there or assigns to a thread in an existing thread pool.  The threadworker can also manage connections and update signals.  The code then looks like:

```
def run_calibration(self):
    self.calib_worker = CalibrationWorker(self, self.calib)
    QThreadWorker.start_thread(self.calib_worker)             # here, the worker will move
    QThreadWorker.connect_signals(self, self.calib_worker)      # connects plugin_widget elements and calib_worker elements through a call to calib_worker.
    ...

```

And this allows you to repeat the same structure for all subsequent threads.

Additionally, you often have an `update` and `finished` property for the "worker" classes.  This could become an interface or similar.  Also, some other actions like `worker.deleteLater` and `thread.deleteLater` can all be handled from the QThreadWorker class.

Reply:
Yeah I have to think about this a bit-- I do think that the code to start / manage threads needs to be consolidated